File: test/performance/data-structures/binary-tree/tree-multi-set.test.cpp
#include <benchmark/benchmark.h>
#include <set>
#include <vector>
#include <random>

static const int MILLION = 1000000;
static const int HUNDRED_THOUSAND = 100000;

// Global RNG
static std::mt19937 gen(42);

static std::vector<int> generateRandomArray(int size, int max_value) {
  std::vector<int> arr;
  arr.reserve(size);
  std::uniform_int_distribution<> dist(0, max_value - 1);
  for (int i = 0; i < size; i++) arr.push_back(dist(gen));
  return arr;
}

// Use fewer unique keys to create duplicates (multiset scenario)
static const auto randomKeys = generateRandomArray(MILLION, HUNDRED_THOUSAND);

// Prebuilt multiset for lookup-only benchmarks
static std::multiset<int> prebuiltMultiSet;
static bool multiSetInitialized = false;

static void initializeMultiSet() {
  if (!multiSetInitialized) {
    for (int i = 0; i < MILLION; i++) {
      prebuiltMultiSet.insert(randomKeys[i]);
    }
    multiSetInitialized = true;
  }
}

// -----------------
// ADD (insert with duplicates)
// -----------------
static void Add1M(benchmark::State& state) {
  for (auto _ : state) {
    std::multiset<int> ms;
    for (int i = 0; i < MILLION; i++) {
      ms.insert(randomKeys[i]);
    }
    benchmark::DoNotOptimize(ms);
  }
}
BENCHMARK(Add1M)->Name("1M add std::multiset")->Unit(benchmark::kMillisecond);

// -----------------
// HAS (count > 0)
// -----------------
static void Has1M(benchmark::State& state) {
  initializeMultiSet();
  for (auto _ : state) {
    int found = 0;
    for (int i = 0; i < MILLION; i++) {
      if (prebuiltMultiSet.count(randomKeys[i]) > 0) found++;
    }
    benchmark::DoNotOptimize(found);
  }
}
BENCHMARK(Has1M)->Name("1M has std::multiset")->Unit(benchmark::kMillisecond);

// -----------------
// COUNT
// -----------------
static void Count1M(benchmark::State& state) {
  initializeMultiSet();
  for (auto _ : state) {
    size_t total = 0;
    for (int i = 0; i < MILLION; i++) {
      total += prebuiltMultiSet.count(randomKeys[i]);
    }
    benchmark::DoNotOptimize(total);
  }
}
BENCHMARK(Count1M)->Name("1M count std::multiset")->Unit(benchmark::kMillisecond);

// -----------------
// BUILD + HAS
// -----------------
static void BuildHas1M(benchmark::State& state) {
  for (auto _ : state) {
    std::multiset<int> ms;
    for (int i = 0; i < MILLION; i++) {
      ms.insert(randomKeys[i]);
    }
    int found = 0;
    for (int i = 0; i < MILLION; i++) {
      if (ms.count(randomKeys[i]) > 0) found++;
    }
    benchmark::DoNotOptimize(found);
  }
}
BENCHMARK(BuildHas1M)->Name("1M build+has std::multiset")->Unit(benchmark::kMillisecond);

// -----------------
// BUILD + COUNT
// -----------------
static void BuildCount1M(benchmark::State& state) {
  for (auto _ : state) {
    std::multiset<int> ms;
    for (int i = 0; i < MILLION; i++) {
      ms.insert(randomKeys[i]);
    }
    size_t total = 0;
    for (int i = 0; i < MILLION; i++) {
      total += ms.count(randomKeys[i]);
    }
    benchmark::DoNotOptimize(total);
  }
}
BENCHMARK(BuildCount1M)->Name("1M build+count std::multiset")->Unit(benchmark::kMillisecond);

// -----------------
// DELETE ONE (erase single occurrence)
// -----------------
static void DeleteOne100K(benchmark::State& state) {
  for (auto _ : state) {
    std::multiset<int> ms;
    for (int i = 0; i < HUNDRED_THOUSAND; i++) {
      ms.insert(randomKeys[i]);
    }
    for (int i = 0; i < HUNDRED_THOUSAND; i++) {
      auto it = ms.find(randomKeys[i]);
      if (it != ms.end()) ms.erase(it);
    }
    benchmark::DoNotOptimize(ms);
  }
}
BENCHMARK(DeleteOne100K)->Name("100K delete-one std::multiset")->Unit(benchmark::kMillisecond);

// -----------------
// FIRST/LAST
// -----------------
static void FirstLast100K(benchmark::State& state) {
  initializeMultiSet();
  for (auto _ : state) {
    int sum = 0;
    for (int i = 0; i < HUNDRED_THOUSAND; i++) {
      auto first = *prebuiltMultiSet.begin();
      auto last = *prebuiltMultiSet.rbegin();
      sum += first + last;
    }
    benchmark::DoNotOptimize(sum);
  }
}
BENCHMARK(FirstLast100K)->Name("100K first/last std::multiset")->Unit(benchmark::kMillisecond);

// -----------------
// CEILING/FLOOR (lower_bound/upper_bound)
// -----------------
static void CeilingFloor100K(benchmark::State& state) {
  initializeMultiSet();
  for (auto _ : state) {
    int sum = 0;
    for (int i = 0; i < HUNDRED_THOUSAND; i++) {
      // ceiling = lower_bound
      auto ceiling = prebuiltMultiSet.lower_bound(i);
      // floor = --upper_bound (if not begin)
      auto floor = prebuiltMultiSet.upper_bound(i);
      if (floor != prebuiltMultiSet.begin()) --floor;
      
      if (ceiling != prebuiltMultiSet.end()) sum += *ceiling;
      sum += *floor;
    }
    benchmark::DoNotOptimize(sum);
  }
}
BENCHMARK(CeilingFloor100K)->Name("100K ceiling/floor std::multiset")->Unit(benchmark::kMillisecond);

// -----------------
// ITERATION (expanded - all elements)
// -----------------
static void Iterate1M(benchmark::State& state) {
  initializeMultiSet();
  for (auto _ : state) {
    int sum = 0;
    for (const auto& v : prebuiltMultiSet) {
      sum += v;
    }
    benchmark::DoNotOptimize(sum);
  }
}
BENCHMARK(Iterate1M)->Name("1M iterate std::multiset")->Unit(benchmark::kMillisecond);

// -----------------
// SIZE
// -----------------
static void Size1M(benchmark::State& state) {
  initializeMultiSet();
  for (auto _ : state) {
    size_t s = prebuiltMultiSet.size();
    benchmark::DoNotOptimize(s);
  }
}
BENCHMARK(Size1M)->Name("1M size std::multiset")->Unit(benchmark::kMillisecond);

// -----------------
// DISTINCT SIZE (unique element count - need to iterate)
// -----------------
static void DistinctSize1M(benchmark::State& state) {
  initializeMultiSet();
  for (auto _ : state) {
    size_t distinct = 0;
    auto it = prebuiltMultiSet.begin();
    while (it != prebuiltMultiSet.end()) {
      distinct++;
      it = prebuiltMultiSet.upper_bound(*it);
    }
    benchmark::DoNotOptimize(distinct);
  }
}
BENCHMARK(DistinctSize1M)->Name("1M distinctSize std::multiset")->Unit(benchmark::kMillisecond);

BENCHMARK_MAIN();

File: test/performance/data-structures/binary-tree/tree-multi-set.test.mjs
import { TreeMultiSet } from '../../../../dist/esm/index.mjs';
import Benchmark from 'benchmark';
import { getRandomIntArray, magnitude } from '../../../utils/perf.mjs';

const suite = new Benchmark.Suite();

const { HUNDRED_THOUSAND, MILLION } = magnitude;

// MultiSet operations: add/delete/setCount
const N_ADD = MILLION;
const N_QUERY = MILLION;
const N_MUTATE = HUNDRED_THOUSAND;

const randomAdd = getRandomIntArray(N_ADD, 0, N_ADD / 10, false); // Allow duplicates (MultiSet scenario)
const randomQuery = getRandomIntArray(N_QUERY, 0, N_ADD / 10, false);

// Prebuilt for lookup-only benchmarks
const tmsPrebuilt = new TreeMultiSet();
for (let i = 0; i < randomAdd.length; i++) {
  tmsPrebuilt.add(randomAdd[i]);
}

suite
  // Add benchmarks: fresh container inside each run
  .add(`${N_ADD.toLocaleString()} add (TreeMultiSet, expanded iteration)`, function () {
    const tms = new TreeMultiSet();
    for (let i = 0; i < randomAdd.length; i++) tms.add(randomAdd[i]);
    this.val = tms;
  })

  // Lookup-only benchmarks (no rebuild)
  .add(`${N_QUERY.toLocaleString()} has-only (TreeMultiSet)`, function () {
    for (let i = 0; i < randomQuery.length; i++) tmsPrebuilt.has(randomQuery[i]);
  })
  .add(`${N_QUERY.toLocaleString()} count-only (TreeMultiSet)`, function () {
    for (let i = 0; i < randomQuery.length; i++) tmsPrebuilt.count(randomQuery[i]);
  })

  // Build + lookup
  .add(`${N_QUERY.toLocaleString()} build+has (TreeMultiSet)`, function () {
    const tms = new TreeMultiSet();
    for (let i = 0; i < randomAdd.length; i++) tms.add(randomAdd[i]);
    for (let i = 0; i < randomQuery.length; i++) tms.has(randomQuery[i]);
    this.val = tms;
  })
  .add(`${N_QUERY.toLocaleString()} build+count (TreeMultiSet)`, function () {
    const tms = new TreeMultiSet();
    for (let i = 0; i < randomAdd.length; i++) tms.add(randomAdd[i]);
    for (let i = 0; i < randomQuery.length; i++) tms.count(randomQuery[i]);
    this.val = tms;
  })

  // Mutation: delete one occurrence
  .add(`${N_MUTATE.toLocaleString()} delete-one (TreeMultiSet)`, function () {
    const tms = new TreeMultiSet();
    for (let i = 0; i < randomAdd.length; i++) tms.add(randomAdd[i]);
    for (let i = 0; i < N_MUTATE; i++) tms.delete(randomQuery[i]);
    this.val = tms;
  })

  // setCount operation
  .add(`${N_MUTATE.toLocaleString()} setCount (TreeMultiSet)`, function () {
    const tms = new TreeMultiSet();
    for (let i = 0; i < randomAdd.length; i++) tms.add(randomAdd[i]);
    for (let i = 0; i < N_MUTATE; i++) tms.setCount(randomQuery[i], 5);
    this.val = tms;
  })

  // Iteration: expanded (default)
  .add(`${N_ADD.toLocaleString()} expanded iteration (TreeMultiSet)`, function () {
    let sum = 0;
    for (const val of tmsPrebuilt) sum += val;
    this.val = sum;
  })

  // Iteration: entries view
  .add(`${N_ADD.toLocaleString()} entries view (TreeMultiSet)`, function () {
    let sum = 0;
    for (const [val, count] of tmsPrebuilt.entries()) sum += val * count;
    this.val = sum;
  })

  // Size operations
  .add(`${N_ADD.toLocaleString()} size property (TreeMultiSet)`, function () {
    this.val = tmsPrebuilt.size;
  })
  .add(`${N_ADD.toLocaleString()} distinctSize property (TreeMultiSet)`, function () {
    this.val = tmsPrebuilt.distinctSize;
  });

export { suite };
